### ISSUE TITLE
AVX512 and Vec512

### ISSUE THREAD

**Issue by limo1996** [pytorch/pytorch#56187]
## 🚀 Feature

Introduce Vec512 into pytorch to support AVX512

## Motivation

In this issue I would like to discuss introduction of 512b vectors into ATen backend. In the private PyTorch fork of the company I am working for I successfully introduced and tested them (not for x86 architecture though). Therefore if there is a general consensus that Vec512 is needed I can try to add them. 

## Pitch

Mirror implementation of `Vec256`. i.e. add `aten/src/ATen/cpu/vec512` folder with base class `vec512_base.h` which will implement scalar version of the ops. Then I would add classes which vectorize base class for the specific vector element type i.e. `vec512_double` etc. I would also add unifying header file `vec.h` with `VecType<>` aliasing either `Vec256` or `Vec512` depending on the cpu capability. Lastly I would replace all usages of `Vec256` with generic `VecType` and double check that all kernels (and tests) using it rely on the vector size and don't have it hardcoded.

Let me know what you guys think! I am open to any ideas :)

## Alternatives

As far as I know there is no way of using 512b vectors inside ATen backend right now.

**Comment by limo1996** [comment]
@ezyang

**Comment by ezyang** [comment]
We're not fundamentally opposed to AVX512, but it will be difficult to determine when we can actually safely turn it on because on Intel it limits max turbo frequency and so can be a pessimization in some cases; additionally, sometimes the AVX512 FLOPS are about the same as AVX2 and so not an improvement either. That being said, I don't think it's a big problem if we start adding infrastructure for it, so yes, I think we'd accept patches for this.

cc @ngimel @malfet @VitalyFedyunin @dzhulgakov

**Comment by limo1996** [comment]
Cool! I have 2 questions regarding the process..
Should I work on my fork or have a branch here?
Should I create one huge pull request in the end or do it incrementally? 
Thanks.

**Comment by ezyang** [comment]
Incremental is better; you could use https://github.com/ezyang/ghstack/ (I may need to give you perms to push branches). Testing strategy is going to be the first and foremost question reviewers will have.

**Comment by limo1996** [comment]
Ok. I will try to use ghstack for managing multiple depending PRs.
You mean permit to push branches into pytorch right? Yes, can you please give me that perm so I can start working on it. Thanks.
Regarding testing.. Unit tests are trivial since they already exist for Vec256 and they are generic enough to handle Vec512 if tweaked a bit. The more problematic will be to test kernels with Vec512 usage.. Here we can test that each kernel should produce the same output both for Vec256 and Vec512.. If there is some test suite for kernels then we don't need to do it.

**Comment by ezyang** [comment]
Open a dummy PR to pytorch repo first so CLA bot can ask you for a CLA, then I can give you rights.

> Here we can test that each kernel should produce the same output both for Vec256 and Vec512.. If there is some test suite for kernels then we don't need to do it.

We have test suite for kernels but they're not going to test AVX512 by default without prodding.

**Comment by limo1996** [comment]
Done

**Comment by VitalyFedyunin** [comment]
If you already have some code, could you please make DRAFT PR and CC me into it, so we can provide you with early feedback. Ideally, we would love to have vec256 become vec<x> with the support of various architectures.

**Comment by limo1996** [comment]
My idea was to have unifying header for vec256 and vec512 so in the end only vec is exposed.
Something like: 
```
#if defined(CPU_CAPABILITY_AVX512)
template <typename T>
using vec = typename at::vec512::Vec512<T>;
#elif defined(CPU_CAPABILITY_AVX2)
template <typename T>
using vec = typename at::vec256::Vec256<T>;
#endif 
```
where vec512 would implement the same fuctionality as vec256. What do you think?

I will push some code as soon as my company approves CLA..

**Comment by imaginary-person** [comment]
Hello @ezyang, even I wanted to work on it a few weeks back, but I searched old issues with strings such as `AVX512` & `vectorization`, and found an issue which was closed stating PyTorch already supports AVX512, so I thought AVX512 vectorization was somehow supported implicitly. That's why I thought as a vectorization project, I could only enable `float16` on AVX2 (and AVX512 implicitly). 😞

Enabling AVX512 isn't very time-consuming, as in most places, one can simply replace `256` with `512` & `32` with `64`. For BFloat16, though, some newer intrinsics can potentially be used.

May I assist with this initiative? I'm very excited about it!

> We're not fundamentally opposed to AVX512, but it will be difficult to determine when we can actually safely turn it on because on Intel it limits max turbo frequency and so can be a pessimization in some cases; additionally, sometimes the AVX512 FLOPS are about the same as AVX2 and so not an improvement either.

If workloads are mostly AVX512, then it'd indeed be faster.
Also, as you already know, a complete cache line is used in AVX512 computations, so that makes computation more memory-bound than AVX2, so to speak. That's why Intel Ice Lake & Tiger Lake processors have 48 KB L1D caches instead of the _normalized_ 32 KB, so AVX512 is even faster on them! Starting with Intel 11th gen, Intel is even churning out low-end laptop processors with AVX512 support, and even these have 48 KB L1D caches! So, it's probably time to jump on the AVX-512 bandwagon. Intel had scrapped plans for a new Xeon server lineup last year, so I think there aren't any Xeon servers with 48 KB L1D cache per core right now, but I mean AVX512 might run even faster on Xeon servers with 48 KB L1D per core in the future.

Intel HW prefetcher can't prefetch across page boundaries, so determining a _prefetch distance_ & using prefetch compiler builtins (or assembly) can further speed up AVX512 computation in some cases.

**Comment by ezyang** [comment]
>  found an issue which was closed stating PyTorch already supports AVX512, so I thought AVX512 vectorization was somehow supported implicitly

@imaginary-person Huh, that's not my recollection, and is not consistent with #26109 and #4825 which are two older issues on the subject.

> May I assist with this initiative? I'm very excited about it!

Sure! Though I suppose you and @limo1996 will have to figure out how to coordinate patches.

**Comment by imaginary-person** [comment]
> @imaginary-person Huh, that's not my recollection, and is not consistent with #26109 and #4825 which are two older issues on the subject.

Ah, you're right! I had misread #51466. I missed an opportunity there 😞

> Sure! Though I suppose you and @limo1996 will have to figure out how to coordinate patches.

Thank you! :)

**Comment by limo1996** [comment]
@imaginary-person I am working on a draft PR. Will let you know once it's ready and we can start to iterate on it. Thanks!

**Comment by VitalyFedyunin** [comment]
We do use AVX512 when OneDNN kernels are called. However all point-wise operators (and everything else based on Vec256) are still on AVX2. 

Bare in mind the fact of adding AVX512 kernels will require us either making it separate compile option or increasing binary size to ship 4 types of kernels simultaneously (with properly updated dispatch mechanics)

**Comment by limo1996** [comment]
I think having some option like `USE_VEC512` would be a good start.. Later on we can try to ship it simultaneously.

**Comment by ezyang** [comment]
@malfet suggests that we might want to just kill AVX (no suffix), since most reasonable users will have AVX2.

**Comment by limo1996** [comment]
Makes sense to me. I currently don't see a reason for having `CPU_CAPABILITY_AVX` as it's always used in `or` with `CPU_CAPABILITY_AVX2`.

**Comment by imaginary-person** [comment]
@ezyang, I remember someone having posted an issue last month, in which they mentioned they were using an old Xeon server for production, that didn't have AVX2! There might be more such users, so maybe a deprecation notice can be provided for the benefit of such users now, perhaps via a sticky post, and then a brief mention in the release notes upon release?

**Comment by ezyang** [comment]
#54684, you mean?

TBH, the user in question didn't even care about perf, it was a reproducibility problem. But sure, I made a post to https://dev-discuss.pytorch.org/t/dropping-avx-support-avx2-only/202

**Comment by imaginary-person** [comment]
Thank you, @ezyang! That's the one. They weren't concerned about performance, but they mentioned that the server not having AVX2 was of a cloud computing platform. So, I thought such users might suddenly encounter perf problems with a newer PyTorch release, if they'd continue using old machines.

**Comment by imaginary-person** [comment]
>  Ideally, we would love to have vec256 become vec with the support of various architectures.

@VitalyFedyunin @limo1996 

I think that a cleaner way to have a single namespace `vec` is:

1. Creating `aten/src/ATen/cpu/vec` with subdirectories `vec256` and `vec512`.
2. Renaming the `vec256` namespace to `vec`. The files in `vec512` directory would also have `vec` namespace.
3. Renaming `Vec256` (and `Vec512`) to a common name, say `Vec`, and then replacing `Vec256` with `Vec` everywhere in the current codebase.
4. The directory `aten/src/ATen/cpu/vec` would have `vec.h` and `functional.h` files, apart from subdirectories `vec256` and `vec512`.
For example, if AVX512 is to be enabled, `vec.h` can simply have the following content:
```
#if defined(CPU_CAPABILITY_AVX512)
#include <ATen/cpu/vec512/vec512.h>
#else
#include <ATen/cpu/vec/vec256/vec256.h>
```
5. `vec.h` should be included wherever `vec256.h` has been included (apart from `vec.h`, and the files in the `vec256` subdirectory, of course).

Please share your opinion. Thanks!

**Comment by imaginary-person** [comment]
> Bare in mind the fact of adding AVX512 kernels will require us either making it separate compile option or increasing binary size to ship 4 types of kernels simultaneously (with properly updated dispatch mechanics)

As of now, my suggestion would only work for the separate compile option. 😞

**Comment by imaginary-person** [comment]
Please confirm which AVX-512 instructions sets should be initially supported. For instance, Caffe2 Perfkernels seem to use `AVX512F`, `AVX512VL`, and `AVX512DQ` instruction sets.
The latest Intel Xeon server I've access to is actually an old one from Q3'17 (Intel Xeon Gold 6142), which supports `AVX512F`, `AVX512BW`, `AVX512CD`, `AVX512DQ`, `AVX512VL`.

However, I don't have access to newer Intel machines that support newer AVX512 instruction sets, such as the `AVX512-VBMI` category (eg. [`_mm512_permutex2var_epi8`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm512_permutex2var_epi8)). Should we create some CPU capability flags for more recent AVX-512 instruction sets such as `AVX512-VBMI` as well, and use those flags to conditionally use different intrinsics, wherever appropriate? Thank you!

**Comment by imaginary-person** [comment]
Hey @limo1996, I added `vec512` in #56992 (It's working locally but might take me some time to get things sorted on the CI infrastructure). It compiles with AVX512 instruction sets `AVX512F`, `AVX512BW`, `AVX512DQ`, and `AVX512VL`. Maybe we can combine this draft PR with the one you'd submit.

Right now, I'd like to focus on correctness, before focusing on optimization.

Hey @ezyang,
> Testing strategy is going to be the first and foremost question reviewers will have.

> We have test suite for kernels but they're not going to test AVX512 by default without prodding.

Can you please clarify which test suite you're referring to? Thanks!
I'll modify `vec256_test_all_types.cpp` to test AVX512 as well & will rename this file to `vec_test_all_types.cpp`.
Also, since both AVX2 & AVX512 use FMA, I was thinking of matching their respective results.
Any advice is most welcome! :)
Thank you!

**Comment by imaginary-person** [comment]
Hello @limo1996, can you please confirm if your company has permitted you to contribute to PyTorch (the ATen AVX512 implementation)? 

Since it's been more than a week, is there a possibility that they might not approve? If so, do you think they'd allow you to at least contribute on incremental revisions?

Thanks!

**Comment by imaginary-person** [comment]
@ezyang @VitalyFedyunin @limo1996 

I think what might make such a PR easier to review for reviewers is annotated comments on the AVX512 intrinsics used that don't have corresponding AVX2 intrinsics.

**Comment by imaginary-person** [comment]
From @apaszke's [old comment](https://github.com/pytorch/pytorch/issues/4825#issuecomment-360159515), it looks like we might have to test on machines with GPUs as well. Perhaps we can run benchmarks to address this concern.

I'd complete testing the implementation today, as I'd have access to a Xeon Silver 4114 machine with an Nvidia P100 GPU only tomorrow, before May 7 (On & after May 7, I can run @limo1996's code on those machines, if his company would permit him to contribute). 
As for testing on a AVX512 supporting laptop with a GPU , I can try to get my hands on one.

**Comment by limo1996** [comment]
@imaginary-person I have it ready (there are todos for some intrinsics) but haven't received approval yet :((

**Comment by imaginary-person** [comment]
Thanks for the update, @limo1996!
Maybe your company sees it as helping the competition.

**Comment by limo1996** [comment]
@imaginary-person I am trying my best to get it..

**Comment by imaginary-person** [comment]
>  (there are todos for some intrinsics)

@limo1996, can you please merge my implementation locally for those TODOs, if possible? Thanks!

**Comment by imaginary-person** [comment]
Thanks to the PyTorch build process, the code-organization I suggested in a [previous comment](https://github.com/pytorch/pytorch/issues/56187#issuecomment-827314981) produces binaries with all vectorization levels - viz. DEFAULT(no vectorization), AVX2, and AVX512 (since I removed AVX)!

**Comment by imaginary-person** [comment]
Hey @limo1996, I'm failing these 5 tests for AVX512 after further progress today:

```
[  PASSED  ] 252 tests.
[  FAILED  ] 5 tests, listed below:
[  FAILED  ] QuantizationTests/0.Quantize, for qint8
[  FAILED  ] QuantizationTests/0.ReQuantizeFromInt, for qint8>
[  FAILED  ] QuantizationTests/1.Quantize, for quint8
[  FAILED  ] QuantizationTests/1.ReQuantizeFromInt, for quint8
[  FAILED  ] QuantizationTests/1.WideningSubtract, for quint8>
```
Please let me know whether they're passing at your end. In that case, I wouldn't read up on quantization in order to debug these. Thanks!

**Comment by limo1996** [comment]
@imaginary-person does `_mm512_div_epi64` compiles for you? For me GCC complains that such function is not defined in intrinsic header file but I can see it here: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm512_div_epi64&expand=5236,2130 

Regarding tests I replaced Vec256 with Vec512 in `vec256_test_all_types.cpp`. I think they are good enough as unit tests. I just don't know how to test kernels..

**Comment by imaginary-person** [comment]
Thanks for your response, @limo1996.

> does `_mm512_div_epi64` compiles for you? For me GCC complains that such function is not defined in intrinsic header file

Yes, it works for me. Which gcc version are you using, BTW? Some intrinsics are not present in older gcc versions.

> Regarding tests I replaced Vec256 with Vec512 in `vec256_test_all_types.cpp`.

Just FYI, one of the tests there has hardcoded inputs for 256 bit vectors.

**Comment by limo1996** [comment]
> Yes, it works for me. Which gcc version are you using, BTW? Some intrinsics are not present in older gcc versions.
Hmm. Ok I will try to update my gcc version and see..

> Just FYI, one of the tests there has hardcoded inputs for 256 bit vectors.
Yeah I know. I hardcoded 512b vectors for now but this test needs to adjust to different vector lengths.

**Comment by imaginary-person** [comment]
> I hardcoded 512b vectors for now

I did so too! . I created another one with an `#if defined(CPU_CAPABILITY_AVX512` :)
(as I'm using a unified namespace `vec` for both `vec256` & `vec512`, and `Vec256` & `Vec512` have a common name). How about using a unified namespace, BTW, as @VitalyFedyunin suggested?

**Comment by limo1996** [comment]
Yeah I am using `vec` as unified namespace and `VecType` as currently used vector type.. What is your gcc version because even after update `_mm512_div_epi64` is not found for me..

**Comment by imaginary-person** [comment]
> What is your gcc version

10.2

**Comment by imaginary-person** [comment]
> Yeah I am using `vec` as unified namespace and `VecType` as currently used vector type.. What is your gcc version because even after update `_mm512_div_epi64` is not found for me..



> even after update `_mm512_div_epi64` is not found for me..

Sorry, I was on my phone earlier. I just checked & am not using that intrinsic anywhere. Sorry for the confusion

**Comment by imaginary-person** [comment]
> even after update `_mm512_div_epi64` is not found for me..

Sorry for the confusion, @limo1996! I was using github.com on my phone earlier. I just checked & am not using that intrinsic anywhere. Where do you need to use it, BTW?

I'm using `mm512_div_pd`, though. 
BTW, I just pushed updated code from yesterday.

**Comment by limo1996** [comment]
Yeah it doesn't work for gcc 10.2 either..

**Comment by imaginary-person** [comment]
> Yeah it doesn't work for gcc 10.2 either..

It wouldn't work on [any available gcc version](https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/avx512fintrin.h). Where are you using it, BTW?

**Comment by imaginary-person** [comment]
> we might want to just kill AVX (no suffix), since most reasonable users will have AVX2.

@malfet @ezyang, 

Apart from `ATen`, should `caffe2_perfkernels` retain AVX support?
Thanks!

**Comment by imaginary-person** [comment]
BTW @limo1996, have you tried building locally on Windows (with Visual Studio), and on MacOS?
I haven't, but CI builds fail for them, so I guess some additional changes would be required to be able to build in those environments.

**Comment by ezyang** [comment]
> We'd disable AVX support in ATen, but should caffe2_perfkernels still retain AVX support?

Nyeh, it doesn't matter. Kill it if it makes your life easier, leave it alone if that's easier.

**Comment by imaginary-person** [comment]
> [  FAILED  ] QuantizationTests/0.Quantize, for qint8
> [  FAILED  ] QuantizationTests/0.ReQuantizeFromInt, for qint8>
> [  FAILED  ] QuantizationTests/1.Quantize, for quint8
> [  FAILED  ] QuantizationTests/1.ReQuantizeFromInt, for quint8



Hey @limo1996, sorry for bothering you again, but as I asked earlier, can you please confirm if these 4 tests are passing with the source-code you wrote)?
I'm using gcc 10.2 on a system with Intel Gold 6142 processors (Skylake-SP), on which [`pack_saturate_and_clamp`](https://github.com/pytorch/pytorch/blob/4749bb9eb450b0329796daddfef520309c4fe27e/aten/src/ATen/cpu/vec/vec512/vec512_qint.h#L74-L186) somehow causes a divergence in output (between AVX512 & AVX2) for these test cases (when run with the same seed). 

In case these tests are passing for you, then I won't debug any further.  Thanks!

**Comment by imaginary-person** [comment]
All 257 tests are passing for me now! I've pushed the code if you'd like to test locally.
@ezyang, no tests for `BFloat16` were added when `vec256_bfloat16.h` was initially added to the codebase by Intel folks. 😞 
I'll add a few.

BTW, I got access to an Intel Xeon Silver 4114 machine with an Nvidia GPU sooner than expected, so I'll run benchmarks tomorrow.

**Comment by limo1996** [comment]
@imaginary-person
My company did not give me permission so I will work on it but after working hours. I finally had time to look over your code and I think there are many similarities just overall I am having both `Vec512` and `Vec256` and unifying them in `vec.h` as follows:

```
#if defined(USE_VEC512)
  #include <ATen/cpu/vec512/vec512.h>
  namespace at { namespace vec {
  using at::vec512::maximum;
  using at::vec512::minimum;
  using at::vec512::fmadd;
  using at::vec512::clamp_min;
  using at::vec512::clamp_max;
  using at::vec512::clamp;
  using at::vec512::convert;
  using at::vec512::convert_to_int32;

  using at::vec512::int_same_size_t;
  using at::vec512::cast;
  using at::vec512::mask_gather;
  using at::vec512::gather;
  using at::vec512::convert_to_int_of_same_size;
  using at::vec512::interleave2;
  using at::vec512::deinterleave2;
  }}
  template <typename T>
  using VecType = typename at::vec512::Vec512<T>;
#else
  #include <ATen/cpu/vec256/vec256.h>
  namespace at { namespace vec {
  using at::vec256::maximum;
  using at::vec256::minimum;
  using at::vec256::fmadd;
  using at::vec256::clamp_min;
  using at::vec256::clamp_max;
  using at::vec256::clamp;
  using at::vec256::convert;
  using at::vec256::convert_to_int32;

  using at::vec256::int_same_size_t;
  using at::vec256::cast;
  using at::vec256::mask_gather;
  using at::vec256::gather;
  using at::vec256::convert_to_int_of_same_size;
  using at::vec256::interleave2;
  using at::vec256::deinterleave2;
  }}
  template <typename T>
  using VecType = typename at::vec256::Vec256<T>;
#endif
```
What do you think about that?

I compiled and run your branch and 2 tests are failing. Namely: `ProcessGroupMPITest` and `vec_test_all_types_AVX512`. 
These vector tests fail:
```
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] Arithmetics/2.Multiplication, where TypeParam = at::vec::(anonymous namespace)::Vectorize<c10::complex<float> >
[  FAILED  ] Arithmetics/2.Division, where TypeParam = at::vec::(anonymous namespace)::Vectorize<c10::complex<float> >
[  FAILED  ] Arithmetics/3.Multiplication, where TypeParam = at::vec::(anonymous namespace)::Vectorize<c10::complex<double> >
[  FAILED  ] Arithmetics/3.Division, where TypeParam = at::vec::(anonymous namespace)::Vectorize<c10::complex<double> >
[  FAILED  ] QuantizationTests/0.Quantize, where TypeParam = at::vec::(anonymous namespace)::Vectorize<c10::qint8>
[  FAILED  ] QuantizationTests/1.Quantize, where TypeParam = at::vec::(anonymous namespace)::Vectorize<c10::quint8>
```
For complex tests I see that issue is accuracy respectively no tolerance..

can you give me permision to push to your repo so we can collaborate there? I am going to review your code again more precisely and in case I find something I can create new branch and then PR to `only_vec` branch. Thank you!

**Comment by imaginary-person** [comment]
> I am having both `Vec512` and `Vec256` and unifying them in `vec.h` as follows

Eventually, we'd have to use `USE_AVX512`, just as you did, but as for the rest, it'd be great to get advice from @ezyang, @VitalyFedyunin, and @malfet.

> can you give me permision to push to your repo so we can collaborate there?

Sure! :)
I just sent you an invite.
 
> I compiled and run your branch and 2 tests are failing. Namely: `ProcessGroupMPITest` and `vec_test_all_types_AVX512`.
 These vector tests fail:
 
Thanks for letting me know! I just checked again & it looks like those complex tests are failing due to a regression.
I'll compare older versions in my work repo & push changes.
The last two would've been failing due to a typo that doesn't exist in my local repo, so I copy-pasted that change via the GitHub editor. 

I'll look into `ProcessGroupMPITest`. Thanks!

**Comment by imaginary-person** [comment]
@ezyang @malfet

Currently, the following tests of CI checks are failing for #56992, but I'm unable to reproduce the issues locally on various machines (AVX only, AVX + AVX2, AVX + AVX2 + AVX512), even if I download & install the artifacts of the CI builds for which those CI checks are run, and test them with various `ATEN_CPU_CAPABILITY` values.

FWIW, the CircleCI machine running `pytorch_macos_10_13_py3_test` runs MacOS on an old Xeon E5-2697 server with only AVX-1.0 support.
The rest 3 CI checks seem to be run on machines having AVX512 support.

CI check | Test name | Failure cause
------------------------- | --------------- | ----------------
pytorch_linux_xenial_py3_clang5_asan_test1|test_variant_consistency_eager_std_cpu_complex64  | SIGILL
pytorch_linux_xenial_py3_clang5_asan_test2|test_get_torch_func_signature_exhaustive_std_cpu_float32|SIGILL
pytorch_macos_10_13_py3_test| test_lkj_cholesky_log_prob | Expected value != Actual value
pytorch_linux_bionic_py3_8_gcc9_coverage_test2|LiteInterpreterTest.UpsampleNearest2d|SIGILL


@ezyang, since these issues don't reproduce at my end, is it possible for you to grant me permission to debug these on CircleCI via SSH? Thank you!

**Comment by imaginary-person** [comment]
Hello @peterjc123, would you be able to help with Windows builds in #56992? Thank you!

**Comment by imaginary-person** [comment]
Hello @ezyang, @malfet, @VitalyFedyunin, @limo1996 

Please let me know if this quantization benchmark, based on [one written by](https://github.com/pytorch/pytorch/pull/25202#issue-311119969) @jamesr66a, seems okay:
https://gist.github.com/imaginary-person/111e9ece31a4c754bde3c41936cae496.
The results are also at this link.

On a dual-socket Intel Xeon 6142,  `quint` & `qint8` addition on AVX512 was significantly faster than those for AVX2.
However, for `quint32`, `AVX2` & `AVX512` exhibit similar performance, with AVX512 being slightly faster.

For `float` addition, though, AVX2 was a bit faster than AVX512, when multiple threads were used.
But I hadn't used `taskset` to ensure that distinct physical cores were used.
By default, PyTorch only ensures that `OMP_NUM_THREADS` is equal to the number of physical cores.
It doesn't, however, ensure that the threads are assigned to different physical cores.
A user can do so using `taskset`, etc, so I think single-threaded performance numbers are more relevant here.

As an aside, if I turned off one NUMA node & hyperthreading, then `float addition` became faster than it was earlier.
`quantized` addition, however, became slower than earlier if one NUMA node & hyperthreading were turned off.

I'll try to dig deeper into the reasons for the results. Thanks!

**Comment by ezyang** [comment]
The benchmark looks reasonable to me, but @jamesr66a would know best!

**Comment by jamesr66a** [comment]
Looks good to me! Pretty exciting that AVX512 can make the quantized stuff faster

**Comment by imaginary-person** [comment]
Hello @malfet & @ezyang,

Currently, the Linux CI checks that are tests are run on `large` `resource_class` executors, which are run on machines that don't support AVX512.
To test ATen AVX512 kernels, `xlarge` executors would be required, as these are run on machines that support AVX512.
The number of CI tests running on `xlarge` instances can probably be limited in order to not increase CI costs too much.
Please confirm, if for Linux CI checks, using `xlarge` instances for the CI checks `pytorch_linux_bionic_py3_8_gcc9_coverage_test1` & `pytorch_linux_bionic_py3_8_gcc9_coverage_test2` would suffice, or if there are some other Linux CI checks that should be tested with AVX512 support as well.

BTW, I'm currently left with fixing 2 bugs & enabling AVX512 support for 2 kernels, apart from collating & presenting benchmark results.

Thank you!

**Comment by imaginary-person** [comment]
Since Intel Cascade Lake machines (the only AVX-512 supporting machines available for Linux CI) support the [`AVX512_VNNI`](https://en.wikichip.org/wiki/x86/avx512_vnni) instruction set, we can use them with a flag, wherever appropriate, since they can now be tested on CI!

**Comment by imaginary-person** [comment]
If `normal_fill` (used by `normal_kernel`) is to have `AVX512` vectorization-support & produce the same results as `AVX2`, then it'd be around 5% slower than the current AVX2 implementation for [this benchmark script](https://gist.github.com/imaginary-person/2afa73100befcab65efa61b410f98795). That's because the [`AVX512` implementation](https://github.com/pytorch/pytorch/blob/aa9821fdc2cb346ac6059dbfceff98e25353e730/aten/src/ATen/native/cpu/DistributionTemplates.h#L136-L180) can't leverage spatial-locality the same way as the current `AVX2` implementation. So, I believe enabling AVX512 support for `normal_fill` isn't a good idea.

Please read on for the explanation, if interested.

The Box-Mueller transform used by the `AVX2` implementation is a `2 -> 2` mapping,
and performs computation of the n<sup>th</sup> float in a group of 8 floats with the n<sup>th</sup> float in the next group of 8 floats. This was done to exploit spatial locality for AVX2.

As such, in order to achieve the same results as the `AVX2` implementation (and the unvectorized implementation, which had also been modified to produce the same results as the AVX2 implementation), a `__m512` would have to get 256 bits from every other group of 8 floats.

This involves 8 `memcpy`s (loading & saving) & slows things down.
eg.,
```
  __m512 u1, u2;
  // the first & third groups of 8 floats from 'data' are loaded in u1
  std::memcpy(reinterpret_cast<float*>(&u1), data, 32);
  std::memcpy(reinterpret_cast<float*>(&u1) + 8, data + 16, 32);
  // the second & fourth groups of 8 floats from 'data' are loaded in u2
  std::memcpy(reinterpret_cast<float*>(&u2), data + 8, 32);
  std::memcpy(reinterpret_cast<float*>(&u2) + 8, data + 24, 32);
```

So, I'll revert changes in `avx_mathfun.h` & `DistributionTemplates.h`, which are in `aten/src/ATen/native/cpu`.
Thank you!

**Comment by ngimel** [comment]
Pytorch does not guarantee bitwise reproducibility across different different hardware, so I'm not sure we should strive for exact same (as opposed to statistically valid) results produced by `normal`. Currently, cuda and cpu would produce different samples for "same" seed.

**Comment by imaginary-person** [comment]
Thanks for the clarification, @ngimel!

If I try using groups of 16 floats for the AVX512 implementation, (just as the AVX2 implementation uses groups of 16 floats), the difference in outputs of PyTorch & numpy for some tests exceeds the tolerance by a bit. Please advise if it'd be okay to adjust it.

eg. `test_tensorsolve_cpu_float32`
```
AssertionError: False is not true : Scalars failed to compare as equal! Comparing 14.334077835083008 and 14.333298683166504 gives a difference of 0.0007791519165039062, but the allowed difference with rtol=1.3e-06 and atol=0.0001 is only 0.00011863328828811646!
```

Thank you!

**Comment by ngimel** [comment]
cc @IvanYashchuk, @mruberry for tensorsolve question, the difference does seem a bit large.

**Comment by imaginary-person** [comment]
Surprisingly, I'm unable to reproduce that issue now, but I'm running all tests & would report back once I find the failing test.
Sorry for the inconvenience, @mruberry & @IvanYashchuk, I'll request you for help again if any tests would fail with a large difference with numpy.

BTW, I think `normal_fill` is a good candidate for prefetching. I was able to speed-up both the AVX2 & the AVX512 implementations by prefetching the cache line to be used (it's both read & written to) in the next iteration, even for small tensors!

However, since this operation is so widely used, I should run more than benchmarks than a simple script.
I'll update this comment with their results.
Thanks!

**Comment by imaginary-person** [comment]
Hello @mruberry & @IvanYashchuk,

Surprisingly, `test_sgd` in `test/test_optim.py` [fails](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.2-py3.6-test2/328/console) if I create an [AVX512 implementation of `normal_fill`](https://github.com/pytorch/pytorch/blob/7e8e7795b5f643912651a84361f7d125b2af6100/aten/src/ATen/native/cpu/DistributionTemplates.h#L139-L193) based on the current AVX2 implementation. The failure cause is that the output is less than [`initial_value`](https://github.com/pytorch/pytorch/blob/1d37f415671976c36f6bf3fa4b83384db066fc2f/test/test_optim.py#L118-L127). The optimizer it fails for is `optim_mt.SGD`.

The Box-Mueller transform used by the `AVX2` implementation is a `2 -> 2` mapping, and performs computation of the n<sup>th</sup> float in a group of 8 floats with the n<sup>th</sup> float in the next group of 8 floats. This was done to exploit spatial locality for AVX2. In the failing AVX512 implementation, I perform computation of the n<sup>th</sup> float in a group of 16 floats with the n<sup>th</sup> float in the next group of 16 floats. Can you please confirm if this approach is also statistically valid?

BTW, this error doesn't occur if I create an [AVX512 implementation that delivers the same output as the current AVX2 implementation](https://github.com/pytorch/pytorch/blob/aa9821fdc2cb346ac6059dbfceff98e25353e730/aten/src/ATen/native/cpu/DistributionTemplates.h#L136-L216). However, the issue with this approach is that it's slower than the first approach, as it requires some `memcpy`s. 

Can you please take a look whenever you can?
Thank you!

**Comment by ngimel** [comment]
Yes, it does look like it's exposing underlying flakyness of the test. Given that enabling AVX512 gives just 6% boost, I don't think it makes sense to continue this investigation, so maybe it makes sense to just leave `normal` without AVX512? What do you think, @imaginary-person ?

**Comment by imaginary-person** [comment]
Thanks for your advice, @ngimel! I'll revert `normal` back to its original state.
BTW, the AVX512 performance-boost seems to be quite less than expected, so I'll investigate it at my end, as I might be able to glean some insights that I can apply to optimize other AVX512 kernels.

**Comment by ezyang** [comment]
> BTW, the AVX512 performance-boost seems to be quite less than expected, so I'll investigate it at my end, as I might be able to glean some insights that I can apply to optimize other AVX512 kernels.

This is not that surprising, given throttling :)

**Comment by imaginary-person** [comment]
> This is not that surprising, given throttling :)

It's better for non-floating types, though.

**Comment by imaginary-person** [comment]
Hello @ezyang, the reason behind the low performance boost in `normal` with AVX512 is that the vectorized code in `DistributionTemplates.h` & `avx_mathfun.h` is around 1/8 (results for a tensor sized 10000x10000 had incomplete estimates, so I used a 100000x100000 one, just to profile) of the time spent in `normal_fill_AVX512` or `normal_fill_AVX2`, and hence the speedup in vectorized code doesn't amount to much.

I profile the code with VTune.
I started profiling at the calls (attached VTune to the process at this point) to `normal_fill_AVX512` & `normal_fill_AVX2` & used `exit(0)` to exit the process after return from these functions, so that VTune would only get relevant data. Both of these function calls account for around 5/6 of the time spent in the `normal_` benchmark script, so that means that the vectorized code is an even smaller part of the benchmark script, whose results I had shared earlier.
![normal AVX2](https://user-images.githubusercontent.com/76181208/120567591-cecf2900-c3d7-11eb-8ab5-27d9fdd95c13.PNG)
![normal AVX512](https://user-images.githubusercontent.com/76181208/120567595-d0005600-c3d7-11eb-99fd-2f777f556a24.PNG)

**Comment by ezyang** [comment]
@imaginary-person Doesn't that just mean we need to vectorize the random number generator? ;)

Apropos of nothing, there was a recent internal post about using AVX512-256 (i.e., use the extra registers, but still only AVX256 instructions) and getting a speedup over AVX512 (because of skipping throttling). Might also be something interesting to think about.

**Comment by imaginary-person** [comment]
>  Doesn't that just mean we need to vectorize the random number generator? ;)

Yes, but I'm guessing doing so entails a huge effort, as nobody seems to have brought that up (until now). :sweat:
I haven't gone through its source code yet, though, so I'll see if it's doable for me.

> Apropos of nothing, there was a recent internal post about using AVX512-256 (i.e., use the extra registers, but still only AVX256 instructions) and getting a speedup over AVX512 (because of skipping throttling). Might also be something interesting to think about.

Thanks for the info! That sounds interesting but I wonder if a similar approach can be used to make vanilla AVX512 code faster as well. On Intel server products, at least, the throttling is only a few MHz, and is limited to the specific core being used. Intel is definitely giving more AVX512 kool-aid than it should, but I think AVX512 does work well on Xeon machines for some workloads.

Off the top of my head, I think one such technique, apart from loop-unrolling, is using multiple accumulators where one can. eg. https://stackoverflow.com/a/65827668
Unfortunately, I think such a technique can't be used in `avx_mathfun.h`, but I'll see if it can be used in `vec`.
Thanks for the hint, though! I'll try to think of (and search for) more such techniques.

BTW, a lot of AVX512 use-cases in PyTorch would do better with AVX512-VBMI & AVX512-VBMI2 instruction sets,
but they aren't even available in Intel Cascade Lake machines. Also, many Xeon machines have 2 AVX512 FMA units per core.
So, is it possible that the folks who are doing AVX512-256 might have run the benchmarks on their laptops, and not on a production-grade machine?

**Comment by imaginary-person** [comment]
@ezyang, 

BTW, I think it's important to note that AVX2 causes downclocking as well, and the additional downclocking caused by AVX512 may not hamper performance on some Skylake machines & beyond, because of the double vector-size. I think that this post with credible references is [a must-read](https://community.intel.com/t5/Software-Tuning-Performance/Unexpected-power-vs-cores-profile-for-MKL-kernels-on-modern-Xeon/m-p/1133869/highlight/true#M6450). Basically, AVX512 would _probably not_ hurt performance on a high-end machine, but measurements are recommended.

I did read some horror stories about AVX512 downclocking, but they were for low-end Xeon servers.

This [official data](https://www.intel.com/content/dam/www/public/us/en/documents/specification-updates/xeon-scalable-spec-update.pdf) is for the Intel Skylake family, and the first link helps understand its significance. Cascade Lake & Ice Lake SP Xeon servers are said to be even better when it comes to AVX512 performance.

![avx2 turbo freq](https://user-images.githubusercontent.com/76181208/120591929-97c23d00-c402-11eb-99fb-c5408d592f59.PNG)
![AVX512 turbo frquency](https://user-images.githubusercontent.com/76181208/120591932-98f36a00-c402-11eb-9618-a785c34d8b90.PNG)

**Comment by ezyang** [comment]
> How about having such an option for PyTorch, so that folks with AVX512-enabled laptops may benefit, if AVX512 doesn't work as well on their laptops?

I think it would be very reasonable.

>  How about having such an option for PyTorch, so that folks with AVX512-enabled laptops may benefit, if AVX512 doesn't work as well on their laptops?

This testing was on Skylake. But I definitely agree that as time goes on AVX512 should probably become an unambiguous win. Which is why I'm generally supportive of this effort; unless Intel totally cans AVX512, there will probably be some point in the future when the decision to use it or not is very clear.

**Comment by imaginary-person** [comment]
In order to disable 2 tests only for AVX512 (one is flaky even on AVX2, but its flakiness is currently not visible, and the other only fails on Cascade Lake machines), I'd have to create an API to get the value of the current `ATEN_CPU_CAPABILITY`.
Such an API can also allow users to set `ATEN_CPU_CAPABILITY` in their Python scripts, depending upon their needs.

An easy way to do this is to simply set the `ATEN_CPU_CAPABILITY` environment variable in `get_cpu_capability()`, if it doesn't already exist. But then users would've to import `os.environ` in their Python code to set/get it.

**Comment by ezyang** [comment]
Yes, just test what the capability is and skip if it's not right.

**Comment by imaginary-person** [comment]
@ezyang, regarding `AVX512_256`, when I compile the AVX2 kernels of the current master branch's source-code with `-march=native` instead of `-mavx2`, I don't see registers `ymm16-ymm31` being used for all kernels. So I think gcc/g++ 9.3 doesn't always leverage extra registers & perhaps I'd have to use compiler optimization techniques at various places in the code to use them. As such, I think I should skip optimizing `AVX512_256` for now, and if it doesn't provide a speedup over AVX2 in most cases, then perhaps we can consider dropping it for now.

Thanks!

**Comment by imaginary-person** [comment]
I'm adding a custom build option (environment variable `ATEN_AVX512_256`) for `AVX512_256` instead, which, when `TRUE`, would use 32 ymm registers for `CPU_CAPABILITY_AVX2` (if AVX512 is supported).

BTW, it seems that for some tasks, using AVX512 results in slightly less (estimated) power consumption than AVX2! I'm not sure how reliable those RAPL estimates are, though.

**Comment by ezyang** [comment]
> So I think gcc/g++ 9.3 doesn't always leverage extra registers & perhaps I'd have to use compiler optimization techniques at various places in the code to use them

OK, that's disappointing, but not entirely unexpected

**Comment by imaginary-person** [comment]
> not entirely unexpected

I guess that's why FBGEMM doesn't rely upon gcc/g++ & uses assembly in some kernels.
BTW, [FBGEMM is only using `AVX512_256` for Xeon D processors](https://github.com/pytorch/FBGEMM/commit/2c87645b7b3b2153510aa0cc21464bf882ab02f2), which are said to have poor AVX512 performance.

**Comment by imaginary-person** [comment]
Probably a stupid question, but can we have a fast path for tensor addition, when `alpha` is 1? `fmadd` is being used for vectorized addition, so for adding 2 float tensors, even `CPU_CAPABILITY_DEFAULT` is sometimes slightly faster than `CPU_CAPABILITY_AVX2`, which, in turn, is 6%-8% faster than `CPU_CAPABILITY_AVX512`.

**Comment by ngimel** [comment]
We definitely should! Do you have benchmarks for the difference?

**Comment by imaginary-person** [comment]
Sorry @ngimel, `vec256_fmadd_ps` has the same latency as `vec256_add_ps`, so that's probably why no fast path has been added.

I [added two 64 MB tensors of size (1, 256, 256, 256) each](https://gist.github.com/imaginary-person/e0bbac859f5b6c309d8a638e2629df0d). 
The average of 10000 iterations yields different results each time, and since the potential fast-path (`add`, instead of `fmadd`) initially gave a better result, I was mistaken about it being faster.

I tested on a dual-socket Intel Xeon Gold 6142 (Skylake) with 16 cores on each socket, and two AVX512 FMA units per core.
I used the default setting (32 threads, 64 logical cores, 32 physical cores).

Without vectorization, the benchmark took around 8.34 ms/iteration. Each core's frequency was 3.3 GHz.
With AVX2, it took 8.46 ms/iteration. Each core's average frequency was around 3.05 GHz.
With AVX512, it took 9.09 ms/iteration. Each core's average frequency was around 2.55 GHz.

I'm not sure why both AVX2 & AVX512 are slower than the unvectorized version for float addition.
I'll profile the code.

BTW, some Cascade Lake machines have smaller difference between AVX2 & AVX512 turbo frequencies than Skylake, so I think AVX512 would perform better on them. I can try running this benchmark on the Cascade Lake CI machine, although it might be a low-end machine (`lscpu` doesn't show the model name but it shows a value of 5 for `Stepping`, which ranges from 5 to 7 for Cascade Lake).

**Comment by imaginary-person** [comment]
@ngimel, for a single thread, `AVX2` is around 10% faster than the `default` ATen CPU capability for adding float tensors. I'm not sure what accounts for this smaller-than-expected performance boost. If this bottleneck can be identified, it'd help boost AVX512 performance as well.

The performance difference between `default` & `AVX2` ATen CPU capabilities decreases as the number of threads are increased. At some point, `default` becomes faster or as fast as `AVX2`. `AVX512` follows the same trend.

**Comment by imaginary-person** [comment]
I do realize that there's a lot more going around with tensors in PyTorch, but if I were to create & run a [simple C++ program](https://gist.github.com/imaginary-person/80b749dcc8560ffb7294aa6b84bc7a21) with two 64 MB arrays of floats, fill them up with random values, and then benchmark (with RDTSCP, for accuracy) the time taken for adding them & storing the output in a third array of floats 1000 times, then I get the following timings with a single thread -

Vectorization-type    |     ms/iter (float)          
-----------------------|-------------------------
Unvectorized             | 4.684 ms/iteration
AVX2                         | 0.587 ms/iteration
AVX512                     | 0.300 ms/iteration

This means that there's room for optimizing tensor addition with `ATEN_CPU_CAPABILITY=avx2` (as well as `ATEN_CPU_CAPABILITY=avx512`), compared to `ATEN_CPU_CAPABILITY=default`.

**Comment by imaginary-person** [comment]
Even if I use `torch.add(x, y, out=z)`, instead of `x + y`, `ATEN_CPU_CAPABILITY=avx2` is only 12% faster than `ATEN_CPU_CAPABILITY=default` for a single thread, with `ATEN_CPU_CAPABILITY=avx2` gradually getting worse as the number of threads are increased. 😞

**Comment by imaginary-person** [comment]
@ezyang @ngimel,

Looks like those performance numbers looked poor because three 64 MB tensors made the computation memory-bound, so it couldn't benefit from vectorization. colesbury was very helpful & proved that with smaller tensors, vectorization's benefits can be readily observed. So, I'll use small tensors in custom benchmarks instead & post their results.

BTW, I've reverted `normal_fill` of `normal_kernel` back to its AVX2 implementation, as it exposes flakiness in `test_sgd`, and also, as I found today, in `test_conj_view_linalg_pinv_hermitian`. `CPU_CAPABILITY_AVX512` would use the AVX2 implementation, and as per the Intel Optimization Manual, mixing AVX512 instructions with AVX2 instructions doesn't cause any slowdown.

Thank you!
Have a nice weekend!

**Comment by imaginary-person** [comment]
Hello @ezyang & @colesbury, thank you for your patience!

[This particular benchmark](https://gist.github.com/imaginary-person/3090b8f66801f589732541e34d646522) compares the performance of adding 1 MB float tensors, with adding corresponding quantized tensors. The following text has been copy-pasted from that link, so you might want to skip reading it & visit that link directly.

These results (without graphs) at the link are based on a Xeon Gold 6142 processor. One NUMA node was turned off, and `numactl --memnode=0` was used to ensure that memory attached to the active node/socket was used). `add_kernel` uses FMA & this machine has 2 AVX512 FMA units per core, so I'll also test on a machine with 1 FMA unit per core (and also with different tensor sizes). The number of threads were varied.

Some particular C-states could've been turned off for better accuracy, but they weren't.

### Summary (also at the link)

- When HyperThreading was ON, both Float tensor addition & quantized (quint8, qint8) addition had better performance with AVX512, than with AVX2. 

- With hyperthreading turned ON, quantized tensor addition reached its best timing with 8 threads, while float tensor addition kept getting faster with an increase in the number of threads. 

- With HyperThreading turned off,  float addition was slightly faster with AVX2 for 16 threads, than on AVX512 with 16 threads, while quantized addition was slower with AVX2, than with AVX512. 

- With hyperthreading turned off, Both float addition & quantized tensor addition kept getting faster with an increase in the number of threads.

- The reason for Float tensor addition getting faster than quint8 or qint8 tensor addition at some point (irrespective of whether HyperThreading is ON or OFF) might've something to do with the parallelization overhead becoming a factor for quantized tensors with an increase in the number of threads, but this needs to be verified by varying the tensor size (and by profiling).

I'll post results of more benchmarks, and will create corresponding graphs.
Please also let me know if you've some particular benchmark in mind.
Thank you!

**Comment by imaginary-person** [comment]
### Proposal - giving users the ability to change `ATEN_CPU_CAPABILITY` dynamically.
If users are using a CPU platform that works better for a portion of their tasks with AVX2, and the rest with AVX512,
having an API for changing the `ATEN_CPU_CAPABILITY` dynamically would help. I can create another PR for this, if adding it seems to be a good idea.

cc @colesbury

Assumption: Only one thread reads & writes `ATEN_CPU_CAPABILITY`.

Currently, `DispatchStub.cpp` uses a static variable to store the `CPUCapability`, so that `get_cpu_capability()` is called only once:
https://github.com/pytorch/pytorch/blob/40a7c317bc60713528320b9786765e4ec5707982/aten/src/ATen/native/DispatchStub.cpp#L49-L52


However, if we make the current CPU capability a global variable (as in the following example), then we can also create an API to set `ATEN_CPU_CAPABILITY` dynamically.
```
CPUCapability current_cpu_capability{CPUCapability::DEFAULT};

CPUCapability get_cpu_capability() {
  static std::once_flag get_capability_flag;
  std::call_once(get_capability_flag, []() {
    CPUCapability capability = compute_cpu_capability();
    current_cpu_capability = capability;
  });
  return current_cpu_capability;
}
```

**Comment by ezyang** [comment]
This is probably OK, though not for your stated reason. 

> Assumption: Only one thread reads & writes ATEN_CPU_CAPABILITY. I'm assuming this is correct.
Besides, making reads & writes atomic would slow things down, so if this isn't true, then this idea isn't worth pursuing.

There is only one write to `static CPUCapability capability` but it can be accessed from multiple threads, since multiple threads can be executing kernels that hit DispatchStub. However, this counts as unidirectional data flow, so relaxed access to the std::atomic is OK per Section 2.6.1 of http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2055r0.pdf

**Comment by imaginary-person** [comment]
Thank you, @ezyang, that was very helpful!
But I'm a little confused, so this is probably a stupid question, but the other threads that execute kernels according to DispatchStub don't actually read the value of `CPUCapability`, but only perform computation over their respective chunks, right? This seems to be true for simple kernels such as `add`, but I'm not sure about more involved kernels (that might have nested OpenMP parallel regions?). Do you think it'd be okay to dispense with `std::atomic` (`memory_order_relaxed`) altogether?

BTW, are there any specific benchmarks in the codebase, running whom you think would be beneficial to compare the performance of AVX512 & AVX2? Thank you!

**Comment by imaginary-person** [comment]
For `MaxPool2d`, NCHW is faster with AVX2, but NHWC is faster with AVX512.

https://gist.github.com/imaginary-person/188071a1fb1011c5a3b22d3500f460d9

**Comment by imaginary-person** [comment]
@ezyang @colesbury @ngimel 

Please advise as to which benchmarks I should run, as I'm confused about the ones that'd be the most helpful.
I'm currently re-running `benchmarks/operators/benchmark_all_test.py`, but they seem to  produce slightly different results in each iteration. Moreover, it's huge & hence it's very hard to create graphs for it.
Please also recommend if I should turn off all Processor C states (except C0, the default one).

Thank you!

**Comment by imaginary-person** [comment]
Hello @jamesr66a, can you please give an example of the sort of benchmarks that you think would be informative to compare the performance of AVX2 & AVX512? eg. end-to-end benchmarks, or per-kernel benchmarks. Thanks!

**Comment by imaginary-person** [comment]
@ezyang, would it be okay if I add comments to code wherever I think it'd be helpful for reviewing code?
I'm asking because that might lead to a lot of notifications.
Thanks!

**Comment by imaginary-person** [comment]
Here are some benchmark results from Friday based on https://github.com/pytorch/benchmark, that were taken with Turbo mode enabled. Those benchmarks are meant to be run on Intel Cascade Lake machines (AWS c5d instances) with turbo mode disabled & most processor C states disabled, but I ran them on a Skylake machine, which has lower Turbo frequencies than its Cascade Lake equivalents -

https://gist.github.com/imaginary-person/7e1650e41bdbfc853eaa5e1b609f078d

They were a bit different from the ones I took earlier, so it looks like I'd have to disable Turbo mode after all.

**Comment by dskhudia** [comment]
@imaginary-person,

>  can you please give an example of the sort of benchmarks that you think would be informative to compare the performance of AVX2 & AVX512? eg. end-to-end benchmarks, or per-kernel benchmarks.

**Quantization related:** 
For per-kernel benchmarks, please use what is suggested in the following comment. Also for benchmark result comparison purposes, running with turbo disabled would be a good way to compare. Otherwise, we won't know whether avx512 is generally better/worse or just the frequency scaling is making it so. 
https://github.com/pytorch/pytorch/pull/56992#discussion_r657627648

For full models: Use pre-quantized models from here: https://github.com/pytorch/vision/tree/master/torchvision/models/quantization

The models here have pre-quantized model (e.g., https://github.com/pytorch/vision/blob/master/torchvision/models/quantization/resnet.py#L13-L14) and you can put them in benchmark settings to run them. Again for benchmarking purposes, running with turbo off + single threaded will give us a better idea about isolating performance only due to avx512 and not because of frequency scaling. From these models, running the ones that spend less time in Conv should show greater benefits due to your patch. Conv isn't affected by your patch as it's dispatched to fbgemm kernels. So lesser the overall time in conv, the more the other operators' performance improvement will be visible. 

cc: @jamesr66a, @raghuramank100

**Comment by dskhudia** [comment]
@ezyang, @imaginary-person 
> I guess that's why FBGEMM doesn't rely upon gcc/g++ & uses assembly in some kernels. (Or maybe it's because it uses asm-jit?).
BTW, FBGEMM is only using AVX512_256 for Xeon D processors, which are said to have poor AVX512 performance.

We (FBGEMM) don't rely on compiler for performance critical code (Linear/Conv/EmbeddingLookup) as we want a tighter control over register allocation among other things. However, for some other code written using intrinsics we make loop trip counts a small compile time constant (e.g., https://github.com/pytorch/FBGEMM/blob/master/src/EmbeddingSpMDMAvx512.cc#L332) so that compiler can unroll such loops and use more registers, thus, the cpu can send more requests to memory to saturate bandwidth or reuse the data in those register without reloading. 

Particular Xeon-D we are interested in has the same peak Avx512/Avx2 performance (at fixed frequency) due to the number of Avx512/Avx2 ALUs it has (1 (Avx512) vs 2 (Avx2)). In these cases, we use AVX512_256 and take advantage of extra registers to hold more data in them for reuse while avoiding potential AVX512 throttling.

**Comment by imaginary-person** [comment]
Thanks a lot for your help and for the info, @dskhudia!

I had planned to run the Quantization benchmarks you had suggested earlier, and would also run the ones you suggested today.
I'll also limit the max C state to C1. 

> However, for some other code written using intrinsics we make loop trip counts a small compile time constant (e.g., https://github.com/pytorch/FBGEMM/blob/master/src/EmbeddingSpMDMAvx512.cc#L332) so that compiler can unroll such loops and use more registers

It seems that if ATen kernels are to benefit a bit from `AVX512_256` as well, they too would have to use such a loop-unrolling technique wherever possible.

**Comment by ezyang** [comment]
> But I'm a little confused, so this is probably a stupid question, but the other threads that execute kernels according to DispatchStub don't actually read the value of CPUCapability, but only perform computation over their respective chunks, right? 

Yes. I'm talking about a different situation, which is when you, the USER call operators from different threads, and they each cross to the kernel via DispatchStub. That is when concurrent accesses can happen.

> Do you think it'd be okay to dispense with std::atomic (memory_order_relaxed) altogether?

The primary reason to use std::atomic is so that TSAN doesn't say there is a race when a (benign) race occurs.
